File: app/src/scripts/parse_lanetalk.py
#!/usr/bin/env python3
"""
parse_lanetalk.py — Extract frame-level bowling data from a Lanetalk
"shared session" HTML page into structured JSON.

Source pages look like:
    http://shared.lanetalk.com/<hash>
e.g. http://shared.lanetalk.com/fdd40824dc6635475db40523a6b830781b406df5

Usage:
    # Parse a local HTML file
    python3 parse_lanetalk.py 3_game_JR_20260608.html > out.json

    # Fetch + parse directly from a Lanetalk share URL
    python3 parse_lanetalk.py --url http://shared.lanetalk.com/<hash> > out.json

    # Write to a file instead of stdout
    python3 parse_lanetalk.py input.html -o out.json

Only the Python standard library is used (no bs4/lxml/requests required).

------------------------------------------------------------------------------
How the HTML encodes a frame (the part that matters)
------------------------------------------------------------------------------
Each game is a `Game N` heading followed by ten `<div class="box">N</div>`
frame markers. For every frame the page renders:

  * A pin diagram (`<div class="pins">` containing four `fullGames-row`s of
    4 / 3 / 2 / 1 pins, i.e. the back row first). Each pin <div> carries a
    class that encodes its fate:
        - plain  `fullGames-pin`            -> knocked down on the FIRST ball
        - oval   `fullGames-pin-oval`       -> knocked down on the SECOND ball
        - white  `fullGames-pin-white`      -> still STANDING at end of frame
    The page legend calls black "knocked down with the first throw" and white
    "pins left after the first throw"; verified against the throw counts across
    all sample frames (black == ball-1 pins, oval == ball-2 pins, white == pins
    left standing).

  * The thrown ball values (`<div class="throws">`) as spans:
        - a digit 0-9                       -> pins felled by that ball
        - "X"                               -> strike (10)  (white-on-black box)
        - "/"                               -> spare        (defensive; unseen in sample)
        - "-"                               -> miss / gutter (0)
        - a span with `border: 2px solid red` (red circle) -> SPLIT on that ball

  * The running `<div class="score">` cumulative total for the frame.

The 10th frame wraps its (2-3) balls in `<div class="frames">` with one
`<div class="internal-frame">` per ball, so it has one pin diagram per ball
rather than one per frame.

Pin numbering: the diagram's rows map to the standard USBC pin layout with the
back row first:  row of 4 = pins 7,8,9,10 ; row of 3 = 4,5,6 ; row of 2 = 2,3 ;
row of 1 = 1 (head pin).
"""

import argparse
import json
import re
import sys
import urllib.request


# Diagram rows are rendered back-row-first (4,3,2,1). Map to USBC pin numbers.
PIN_ROWS = [[7, 8, 9, 10], [4, 5, 6], [2, 3], [1]]


def fetch(url: str) -> str:
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=30) as resp:
        return resp.read().decode("utf-8", errors="replace")


def _first(pattern, text, group=1, default=None, flags=0):
    m = re.search(pattern, text, flags)
    return m.group(group).strip() if m else default


def parse_header(html: str) -> dict:
    """Pull the overview / session metadata from the page head + Overview section."""
    title = _first(r'og:title"\s+content="([^"]+)"', html, default="")
    player = _first(r'<div class="user">\s*<h1>([^<]+)</h1>', html, default="")
    datetime_text = _first(r'<div class="second-column">\s*<h2>([^<]+)</h2>', html, default="")
    center = _first(r'class="name">([^<]+)</h2>', html, default="")
    place = _first(r'class="place">([^<]+)</h2>', html, default="")
    if place:
        place = place.lstrip(", ").strip()

    # Overview summary cards: Games / Total / Average
    games = _first(r'<span>Games</span>\s*<h2>(\d+)</h2>', html)
    total = _first(r'<span>Total</span>\s*<h2>(\d+)</h2>', html)
    average = _first(r'<span>Average</span>\s*<h2>(\d+)</h2>', html)

    return {
        "title": title,
        "player": player,
        "bowling_center": {"name": center, "location": place},
        "datetime_text": datetime_text,
        "datetime_iso": _to_iso(datetime_text),
        "summary": {
            "games": int(games) if games else None,
            "total": int(total) if total else None,
            "average": int(average) if average else None,
        },
    }


def _to_iso(text: str):
    """'1:28 AM on Tuesday, June 09, 2026' -> '2026-06-09T01:28:00'."""
    if not text:
        return None
    from datetime import datetime
    cleaned = re.sub(r"^(.*?) on \w+, (.*)$", r"\1 \2", text.strip())
    for fmt in ("%I:%M %p %B %d, %Y", "%I:%M %p %b %d, %Y"):
        try:
            return datetime.strptime(cleaned, fmt).isoformat()
        except ValueError:
            continue
    return None


def parse_pin_diagram(block: str) -> dict:
    """A single pin-diagram segment -> {pin_number: state}.

    `block` is the HTML following a `<div class="pins">` marker (up to the next
    such marker), containing four `fullGames-row` groups of 4/3/2/1 pins.
    state in {"down_first", "down_second", "standing"}.
    """
    rows = re.split(r'class="fullGames-row">', block)[1:]

    state = {}
    for row_idx, row_html in enumerate(rows[:4]):
        pins = re.findall(r'<div class="(fullGames-pin[^"]*)"></div>', row_html)
        numbers = PIN_ROWS[row_idx]
        for pin_idx, cls in enumerate(pins[: len(numbers)]):
            if "fullGames-pin-white" in cls:
                s = "standing"      # filled-white pin = left standing at end of frame
            elif "fullGames-pin-oval" in cls:
                s = "down_second"   # outline pin = knocked down on the second ball
            else:
                s = "down_first"    # solid black pin = knocked down on the first ball
            state[numbers[pin_idx]] = s
    return state


def parse_throws(throws_html: str, prior_pins: int = 0) -> list:
    """A `<div class="throws">` block -> ordered list of ball dicts.

    Each ball: {"display", "pins", "split"} where display is the on-screen
    token (digit / "X" / "/" / "-"), pins is the integer pins felled, and
    split flags the red-circled-count split indicator.

    `prior_pins` is the pins already felled earlier in the same frame; it lets a
    spare (rendered as a `triangle` glyph, not text) resolve to `10 - prior`.
    Tokens are read in document order so a span followed by a triangle yields
    the first ball then the spare.
    """
    balls = []
    running = prior_pins
    # Walk spans and triangle markers in the order they appear.
    for m in re.finditer(r'<span style="([^"]*font-size: 20px[^"]*)">\s*([^<]*?)\s*</span>'
                         r'|<div class="triangle">', throws_html):
        if m.group(0).startswith('<div'):
            # Spare marker: completes the rack relative to what's already down.
            pins = 10 - running
            balls.append({"display": "/", "pins": pins, "split": False})
            running = 10
            continue
        token = (m.group(2) or "").strip()
        if token == "":
            continue
        pins = _token_to_pins(token)
        balls.append({
            "display": token,
            "pins": pins,
            "split": "border: 2px solid red" in m.group(1),
        })
        running += pins or 0
    return balls


def _token_to_pins(token: str):
    if token == "X":
        return 10
    if token == "/":
        return None  # spare value depends on previous ball; caller can resolve
    if token == "-":
        return 0
    if token.isdigit():
        return int(token)
    return None


def parse_frame(frame_html: str, frame_number: int) -> dict:
    """Parse one frame slice (works for frames 1-9 and the 10th)."""
    # All pin diagrams in this frame (1 for frames 1-9, one per ball in the 10th).
    # Each diagram is the segment following a `<div class="pins">` marker; the
    # trailing throws/score markup is ignored since we only read fullGames-rows.
    pin_blocks = re.split(r'<div class="pins">', frame_html)[1:]
    diagrams = [parse_pin_diagram(b) for b in pin_blocks]

    # Throw values across the whole frame (the 10th has several throws blocks).
    # A strike (or in the 10th, completing the rack) resets the running count
    # the spare glyph is measured against.
    balls = []
    running = 0
    for throws_html in re.findall(r'<div class="throws">(.*?)</div>\s*(?:<div class="score"|</div>)',
                                  frame_html, re.S):
        block_balls = parse_throws(throws_html, prior_pins=running)
        balls.extend(block_balls)
        for b in block_balls:
            running = 0 if b["pins"] == 10 else running + (b["pins"] or 0)

    # Cumulative running score for the frame.
    score = _first(r'<div class="score">\s*<span>(\d+)</span>', frame_html)

    is_strike = any(b["display"] == "X" for b in balls)
    # A spare: not a strike, but the first two balls clear all ten pins. Detect
    # via the "/" token or, for frames 1-9, by the first two balls summing to 10.
    is_spare = (not is_strike) and any(b["display"] == "/" for b in balls)
    if not is_strike and not is_spare and frame_number <= 9 and len(balls) >= 2:
        p0, p1 = balls[0]["pins"], balls[1]["pins"]
        if p0 is not None and p1 is not None and p0 + p1 == 10:
            is_spare = True

    return {
        "frame": frame_number,
        "throws": balls,
        "cumulative_score": int(score) if score else None,
        "is_strike": is_strike,
        "is_spare": is_spare,
        "is_split": any(b["split"] for b in balls),
        "pin_diagrams": diagrams,
    }


def parse_game(game_number: int, body: str) -> dict:
    """Slice a game body into its ten frames by the `box` frame markers."""
    box_iter = list(re.finditer(r'<div class="box">\s*<span[^>]*>(\d+)</span>', body))
    frames = []
    for i, m in enumerate(box_iter):
        start = m.start()
        end = box_iter[i + 1].start() if i + 1 < len(box_iter) else len(body)
        frame_number = int(m.group(1))
        frames.append(parse_frame(body[start:end], frame_number))

    final_score = frames[-1]["cumulative_score"] if frames else None
    return {"game_number": game_number, "score": final_score, "frames": frames}


def parse(html: str, source_url: str = None) -> dict:
    result = {"source_url": source_url}
    result.update(parse_header(html))

    # Split into game sections by their headings; restrict to the games container.
    parts = re.split(r'title-headlines">Game (\d+)<', html)
    games = []
    for i in range(1, len(parts), 2):
        game_number = int(parts[i])
        games.append(parse_game(game_number, parts[i + 1]))
    result["games"] = games
    return result


def main(argv=None):
    ap = argparse.ArgumentParser(description="Parse Lanetalk shared-session HTML to JSON.")
    ap.add_argument("input", nargs="?", help="Path to a saved HTML file.")
    ap.add_argument("--url", help="Fetch the HTML from this Lanetalk share URL instead.")
    ap.add_argument("-o", "--output", help="Write JSON here (default: stdout).")
    ap.add_argument("--indent", type=int, default=2, help="JSON indent (default 2).")
    args = ap.parse_args(argv)

    if args.url:
        html = fetch(args.url)
        source_url = args.url
    elif args.input:
        with open(args.input, encoding="utf-8") as fh:
            html = fh.read()
        source_url = None
    else:
        ap.error("provide an input HTML file or --url")

    data = parse(html, source_url=source_url)
    text = json.dumps(data, indent=args.indent, ensure_ascii=False)

    if args.output:
        with open(args.output, "w", encoding="utf-8") as fh:
            fh.write(text + "\n")
    else:
        sys.stdout.write(text + "\n")


if __name__ == "__main__":
    main()

File: supabase/functions/lanetalk-import/parseLanetalk.ts
// parseLanetalk.ts — TypeScript port of app/src/scripts/parse_lanetalk.py.
//
// Extracts frame-level bowling data from a Lanetalk "shared session" HTML page
// (http://shared.lanetalk.com/<hash>) into the same structured shape the Python
// script emits. Kept deliberately close to the Python (same regexes, same field
// names) so the .py remains a readable reference spec. Standard-library only.
//
// The Python is the source of truth for *why* each regex looks the way it does;
// see its module docstring for how the HTML encodes a frame.

// Diagram rows are rendered back-row-first (4,3,2,1). Map to USBC pin numbers.
const PIN_ROWS: number[][] = [[7, 8, 9, 10], [4, 5, 6], [2, 3], [1]]

const MONTHS: Record<string, number> = {
  january: 1, february: 2, march: 3, april: 4, may: 5, june: 6, july: 7,
  august: 8, september: 9, october: 10, november: 11, december: 12,
  jan: 1, feb: 2, mar: 3, apr: 4, jun: 6, jul: 7, aug: 8, sep: 9,
  oct: 10, nov: 11, dec: 12,
}
const MON_ABBR = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

export type PinState = 'down_first' | 'down_second' | 'standing'
export type PinDiagram = Record<string, PinState>

export interface LanetalkThrow {
  display: string
  pins: number | null
  split: boolean
}

export interface LanetalkFrame {
  frame: number
  throws: LanetalkThrow[]
  cumulative_score: number | null
  is_strike: boolean
  is_spare: boolean
  is_split: boolean
  pin_diagrams: PinDiagram[]
}

export interface LanetalkGame {
  game_number: number
  score: number | null
  date: string | null
  date_label: string | null
  played_at: string | null
  source_url: string | null
  frames: LanetalkFrame[]
}

export interface LanetalkSession {
  source_url: string | null
  title: string
  player: string
  bowling_center: { name: string; location: string }
  datetime_text: string
  played_at: string | null
  date: string | null
  date_label: string | null
  summary: { games: number | null; total: number | null; average: number | null }
  games: LanetalkGame[]
}

function first(pattern: string, text: string, group = 1, def: string | null = null, flags = ''): string | null {
  const m = new RegExp(pattern, flags).exec(text)
  return m && m[group] != null ? m[group].trim() : def
}

function pad(n: number): string {
  return String(n).padStart(2, '0')
}

interface ParsedDateTime { year: number; month: number; day: number; iso: string }

/** '1:28 AM on Tuesday, June 09, 2026' -> parts, or null. */
function parseDateTime(text: string): ParsedDateTime | null {
  if (!text) return null
  const cleaned = text.trim().replace(/^(.*?) on \w+, (.*)$/, '$1 $2')
  const m = /^(\d{1,2}):(\d{2})\s*(AM|PM)\s+([A-Za-z]+)\s+(\d{1,2}),\s*(\d{4})$/i.exec(cleaned)
  if (!m) return null
  const month = MONTHS[m[4].toLowerCase()]
  if (!month) return null
  let hour = parseInt(m[1], 10)
  const minute = parseInt(m[2], 10)
  const ampm = m[3].toUpperCase()
  if (ampm === 'PM' && hour !== 12) hour += 12
  if (ampm === 'AM' && hour === 12) hour = 0
  const day = parseInt(m[5], 10)
  const year = parseInt(m[6], 10)
  const iso = `${year}-${pad(month)}-${pad(day)}T${pad(hour)}:${pad(minute)}:00`
  return { year, month, day, iso }
}

/** The Monday of the week a date belongs to (league night). */
function toMonday(dt: ParsedDateTime | null): { date: string; label: string } | null {
  if (!dt) return null
  const d = new Date(Date.UTC(dt.year, dt.month - 1, dt.day))
  const weekday = (d.getUTCDay() + 6) % 7 // Monday == 0
  const monday = new Date(d.getTime() - weekday * 86400000)
  const y = monday.getUTCFullYear()
  const m = monday.getUTCMonth() + 1
  const day = monday.getUTCDate()
  return { date: `${y}-${pad(m)}-${pad(day)}`, label: `${MON_ABBR[m - 1]} ${day}, ${y}` }
}

function parseHeader(html: string) {
  const title = first('og:title"\\s+content="([^"]+)"', html, 1, '') as string
  const player = first('<div class="user">\\s*<h1>([^<]+)</h1>', html, 1, '') as string
  const datetimeText = first('<div class="second-column">\\s*<h2>([^<]+)</h2>', html, 1, '') as string
  const center = first('class="name">([^<]+)</h2>', html, 1, '') as string
  let place = first('class="place">([^<]+)</h2>', html, 1, '') as string
  if (place) place = place.replace(/^[,\s]+/, '').trim()

  const games = first('<span>Games</span>\\s*<h2>(\\d+)</h2>', html)
  const total = first('<span>Total</span>\\s*<h2>(\\d+)</h2>', html)
  const average = first('<span>Average</span>\\s*<h2>(\\d+)</h2>', html)

  const dt = parseDateTime(datetimeText)
  const monday = toMonday(dt)
  return {
    title,
    player,
    bowling_center: { name: center, location: place },
    datetime_text: datetimeText,
    played_at: dt ? dt.iso : null,
    date: monday ? monday.date : null,
    date_label: monday ? monday.label : null,
    summary: {
      games: games ? parseInt(games, 10) : null,
      total: total ? parseInt(total, 10) : null,
      average: average ? parseInt(average, 10) : null,
    },
  }
}

function parsePinDiagram(block: string): PinDiagram {
  const rows = block.split(/class="fullGames-row">/).slice(1)
  const state: PinDiagram = {}
  rows.slice(0, 4).forEach((rowHtml, rowIdx) => {
    const pins = [...rowHtml.matchAll(/<div class="(fullGames-pin[^"]*)"><\/div>/g)].map(m => m[1])
    const numbers = PIN_ROWS[rowIdx]
    pins.slice(0, numbers.length).forEach((cls, pinIdx) => {
      let s: PinState
      if (cls.includes('fullGames-pin-white')) s = 'standing'
      else if (cls.includes('fullGames-pin-oval')) s = 'down_second'
      else s = 'down_first'
      state[numbers[pinIdx]] = s
    })
  })
  return state
}

function tokenToPins(token: string): number | null {
  if (token === 'X') return 10
  if (token === '/') return null
  if (token === '-') return 0
  if (/^\d+$/.test(token)) return parseInt(token, 10)
  return null
}

function parseThrows(throwsHtml: string, priorPins = 0): LanetalkThrow[] {
  const balls: LanetalkThrow[] = []
  let running = priorPins
  const re = /<span style="([^"]*font-size: 20px[^"]*)">\s*([^<]*?)\s*<\/span>|<div class="triangle">/g
  for (const m of throwsHtml.matchAll(re)) {
    if (m[0].startsWith('<div')) {
      const pins = 10 - running
      balls.push({ display: '/', pins, split: false })
      running = 10
      continue
    }
    const token = (m[2] || '').trim()
    if (token === '') continue
    const pins = tokenToPins(token)
    balls.push({ display: token, pins, split: (m[1] || '').includes('border: 2px solid red') })
    running += pins || 0
  }
  return balls
}

function parseFrame(frameHtml: string, frameNumber: number): LanetalkFrame {
  const pinBlocks = frameHtml.split(/<div class="pins">/).slice(1)
  const diagrams = pinBlocks.map(parsePinDiagram)

  const balls: LanetalkThrow[] = []
  let running = 0
  for (const m of frameHtml.matchAll(/<div class="throws">(.*?)<\/div>\s*(?:<div class="score"|<\/div>)/gs)) {
    const blockBalls = parseThrows(m[1], running)
    balls.push(...blockBalls)
    for (const b of blockBalls) running = b.pins === 10 ? 0 : running + (b.pins || 0)
  }

  const score = first('<div class="score">\\s*<span>(\\d+)</span>', frameHtml)
  const isStrike = balls.some(b => b.display === 'X')
  let isSpare = !isStrike && balls.some(b => b.display === '/')
  if (!isStrike && !isSpare && frameNumber <= 9 && balls.length >= 2) {
    const p0 = balls[0].pins
    const p1 = balls[1].pins
    if (p0 != null && p1 != null && p0 + p1 === 10) isSpare = true
  }

  return {
    frame: frameNumber,
    throws: balls,
    cumulative_score: score ? parseInt(score, 10) : null,
    is_strike: isStrike,
    is_spare: isSpare,
    is_split: balls.some(b => b.split),
    pin_diagrams: diagrams,
  }
}

function parseGame(gameNumber: number, body: string): LanetalkGame {
  const boxes = [...body.matchAll(/<div class="box">\s*<span[^>]*>(\d+)<\/span>/g)]
  const frames: LanetalkFrame[] = []
  boxes.forEach((m, i) => {
    const start = m.index!
    const end = i + 1 < boxes.length ? boxes[i + 1].index! : body.length
    frames.push(parseFrame(body.slice(start, end), parseInt(m[1], 10)))
  })
  const finalScore = frames.length ? frames[frames.length - 1].cumulative_score : null
  return { game_number: gameNumber, score: finalScore, date: null, date_label: null, played_at: null, source_url: null, frames }
}

export function parseLanetalk(html: string, sourceUrl: string | null = null): LanetalkSession {
  const header = parseHeader(html)

  // Split into game sections by their headings.
  const parts = html.split(/title-headlines">Game (\d+)</)
  const games: LanetalkGame[] = []
  for (let i = 1; i < parts.length; i += 2) {
    const gameNumber = parseInt(parts[i], 10)
    const game = parseGame(gameNumber, parts[i + 1])
    game.date = header.date
    game.date_label = header.date_label
    game.played_at = header.played_at
    game.source_url = sourceUrl
    games.push(game)
  }

  return { source_url: sourceUrl, ...header, games }
}
